src/sep8/ApprovalProvider.ts

```diff
@@ -0,0 +1,96 @@
+import { Transaction } from "stellar-base";
+import { FeeBumpTransaction } from "stellar-sdk";
+import { ApprovalResponseStatus } from "../constants/sep8";
+import { ApprovalResponse } from "../types/sep8";
+
+export class ApprovalProvider {
+  public approvalServer: string;
+
+  constructor(approvalServer: string) {
+    if (!approvalServer) {
+      throw new Error("Required parameter 'approvalServer' missing!");
+    }
+
+    this.approvalServer = approvalServer.replace(/\/$/, "");
+  }
+
+  public async approve(
+    transaction: Transaction | FeeBumpTransaction,
+    contentType:
+      | "application/json"
+      | "application/x-www-form-urlencoded" = "application/json",
+  ): Promise<ApprovalResponse> {
+    if (!transaction.signatures.length) {
+      throw new Error(
+        "At least one signature is required before submitting for approval.",
+      );
+    }
+
+    const param = {
+      tx: transaction
+        .toEnvelope()
+        .toXDR()
+        .toString("base64"),
+    };
+
+    let response;
+    if (contentType === "application/json") {
+      response = await fetch(this.approvalServer, {
+        method: "POST",
+        body: JSON.stringify(param),
+        headers: {
+          "Content-Type": "application/json",
+        },
+      });
+    } else {
+      const urlParam = new URLSearchParams(param);
+      response = await fetch(this.approvalServer, {
+        method: "POST",
+        body: urlParam.toString(),
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded",
+        },
+      });
+    }
+
+    if (!response.ok) {
+      const responseText = await response.text();
+      try {
+        const { responseJson } = JSON.parse(responseText);
+        throw new Error(
+          `Error sending base64-encoded transaction to ${
+            this.approvalServer
+          }: error ${responseJson.error}`,
+        );
+      } catch (e) {
+        throw new Error(
+          `Error sending base64-encoded transaction to ${
+            this.approvalServer
+          }: error 
+          code ${response.status}, status text: "${responseText}"`,
+        );
+      }
```

Comment:
Do you think we could simplify this a bit? The only thing that is changing is the error message (could we parse JSON or not). Maybe we could set a param in `try catch` and then just use that in the final `throw new Error`? It might be just me, but seems a bit confusing.

Reply:
Looking at it again, I also think it's confusing though this is something we do in other places. Is it correct that if we throw an error at line 60, the new error at line 66 will overwrite the error at line 60?  Meaning once the catch block catches the error at line 60, it will throw the error at line 66.

Reply:
https://github.com/stellar/js-stellar-wallets/pull/203/commits/41ce8e9bfbfa324bc819ce39e39752ce811da566